django/db/models/query.py

```diff
@@ -959,6 +959,8 @@ def reverse(self):
         """
         Reverses the ordering of the QuerySet.
         """
+        if not self.query.can_filter():
+            raise TypeError('Cannot change a query once a slice has been taken')
```

Comment:
I'm not sure about `TypeError` either.  I tried to stay close to similar exceptions like
- [`raise TypeError("Cannot call delete() after .values() or .values_list()")`](https://github.com/django/django/blob/29f6079/django/db/models/query.py#L594)
- [`raise TypeError("Cannot call select_related() after .values() or .values_list()")`](https://github.com/django/django/blob/29f6079/django/db/models/query.py#L856)
- [`raise TypeError("Cannot call defer() after .values() or .values_list()")`](https://github.com/django/django/blob/29f6079/django/db/models/query.py#L975)
- [`raise TypeError("Cannot call only() after .values() or .values_list()")`](https://github.com/django/django/blob/29f6079/django/db/models/query.py#L990)